src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java

```diff
@@ -47,36 +62,52 @@ public void addMessage(Position p) throws InterruptedException {
 			InflightMessage m = new InflightMessage(p);
 			this.linkedMap.put(p, m);
 
-			if (linkedMap.size() >= capacity) {
+			int size = linkedMap.size();
+			if (size >= capacity) {
 				isFull = true;
+			} else if (size == 1) {
+				head = m;
 			}
 		}
 	}
 
 	/* returns the position that stuff is complete up to, or null if there were no changes */
-	public Position completeMessage(Position p) {
+	public InflightMessage completeMessage(Position p) {
 		synchronized (this.linkedMap) {
 			InflightMessage m = this.linkedMap.get(p);
 			assert(m != null);
 
 			m.isComplete = true;
 
-			Position completeUntil = null;
-			Iterator<InflightMessage> iterator = this.linkedMap.values().iterator();
+			InflightMessage completeUntil = null;
+			if (p.equals(head.position)) {
+				Iterator<InflightMessage> iterator = this.linkedMap.values().iterator();
 
-			while ( iterator.hasNext() ) {
-				InflightMessage msg = iterator.next();
-				if ( !msg.isComplete ) {
-					break;
+				while ( iterator.hasNext() ) {
+					InflightMessage msg = iterator.next();
+					if ( !msg.isComplete ) {
+						head = msg;
```

Comment:
You are correct, if all messages are completed in a single go, the `head` isn't reset at the moment. But it can be fixed by having `if (linkedMap.size() == 0)` check.
Without tracking the `head`, we won't have `if (p.equals(head.position))` any more so the 1st `linkedMap.values().iterator()` will run every time. I dislike repeating the code/behaviour that's why having the `head` in the first place. What do you think?